### DISCUSSION TITLE
Load fonts with Non-Latin Characters (for example, Chinese Traditional?)

### DISCUSSION THREAD

**Discussion by ghost** [ocornut/imgui#4974] [New Users Build/Link/Run issues ONLY!]
Here is my "minimal" test case, and I use the word minimal very loosely, as I use a lot of code that is not mine and a lot of code that probably is not necessary for a test case, but it does demonstrate the issue, and if I need to make it way more minimal, I understand and will make a huge effort to reduce the code being used to reproduce with an edit.

Here's what I have, for now:
https://github.com/time-killer-games/filedialogs

in the application you can set an environment variable (this is a Windows-only issue) and if you rename the cloned "fonts" folder to something in non-Latin characters such as Chinese or Cyrillic it will throw runtime errors, and this only happens when loading fonts from paths which contain those characters. I have verified the character are indeed correct and I converted them to use UTF-8 correctly with my widen() and narrow() helper functions.

The code I am using for the directory iterator is a C++17 std::filesystem-like header which can also be used in older (and newer) versions of the C++ standard. I also verified manually that handles UTF-8 correctly in my tests.

The only thing I can at least guess from my testing is at fault here is the lack of UTF-8 in AddFontFromFileTTF() for loading custom fonts from a given file path.

You may use environment_set_variable("IMGUI_FONT_PATH", "/directory/with/fonts/here") and pass a directory containing all the ttf files or otf files you need to load for supporting all languages you need in question for testing.

The runtime error can be tracked with lldb or gdb depending in which you prefer. I have a build script in the repository to make things easier. But I'm pretty sure this is an ImGui issue due to the font error explicitly says it can't find the font even though my file_exists() function check is returning true with IMGUI_FONT_PATH set to a path with Chinese in it where the fonts reside.

**Comment by ghost** [comment]
Here's a screenshot of the error i get:

![Screenshot 2022-02-12 122450](https://user-images.githubusercontent.com/4379204/153711262-af2930d4-35e4-44f1-af94-d3caff2b216b.png)

**Comment by thedmd** [comment]
Are you compiling your program with Unicode support?

See if `_UNICODE` and/or `UNICODE` are defined. These are automatically defined if you choose `Use Unicode Character Set` in project 'Advanced' properties in Visual Studio:
![image](https://user-images.githubusercontent.com/1197433/153754504-8bdb8e30-26b8-45e9-a454-d05d5ce8fede.png)

This should deal with `????` in you project path and maybe this will be enough to get everything work.

**Reply by ghost** [reply]
I'm using MSYS2 / MinGW to build currently, but i can convert my project to a Visual Studio solution if i need to. I've never bothered to explore this possibility but gcc might have a compiler flag on windows that effectively does the same thing as that dropdown box.

**Reply by ghost** [reply]
Thank you for the help, again!

**Reply by thedmd** [reply]
Adding `-D_UNICODE` and `-DUNICODE` might do the trick on MinGW. WinAPI call will be directed to wide functions and they are better at handling non-latin characters. ImGui will use wide function.

If your source file is saved with utf-8 encoding, it will translate properly to wide function.

**Reply by ghost** [reply]
Awesome, thank you! Will try it within a couple hours.

**Reply by ghost** [reply]
![Screenshot 2022-02-15 023528](https://user-images.githubusercontent.com/4379204/153981911-103df77d-532b-4c8a-9162-788a4e3b9cc8.png)

Well at least it's not showing question marks in the error box anymore, so i'll call that much progress.

**Reply by thedmd** [reply]
Please inspect content of `vector<string> fonts` to see if paths computed there are correct. If paths there are wrong this may be an issue with `ghc/filesystem` not ImGui itself.
https://github.com/time-killer-games/ImGuiFileDialog-SDL2-OpenGL2/blob/42257ae653b436d1daaa246f28b3f5889fb2032a/filedialogs.cpp#L394

Also you're searching in working directory by default. Searching in executable directory may be a better idea if your app layout expect `fonts` to be always next to executable.

**Reply by ghost** [reply]
I was cd'ing to the executable directory, because OpenBSD doesn't support getting the executable directory. My code originally used std::filesystem and not ghc::filesystem and iirc both had the same issue. But i can roll back to former version in git history that used std::filesystem and verify this. just to be sure.

**Reply by thedmd** [reply]
You can save executable directory from `argv[0]` in `main()`, which is usually full path to the executable. This is however not a guarantee, this can be empty string too. But is better than nothing. Also it may help if you do not overwrite `IMGUI_FONT_PATH` if it is already present in environment variables.

On Linux, may work on BSD too you can try to get executable path:
```cpp
char result[PATH_MAX];
unsigned int count = (unsigned int)readlink("/proc/self/exe", result, PATH_MAX);
```

All that being said, I think this isn't an ImGui issue.

**Reply by ghost** [reply]
I'll let you know as soon as i have verified that, but it wouldn't surprise me if you were correct. We'll see.

**Reply by ghost** [reply]
I've changed it to use the executable path on all platforms where supported (and am using argv[0] on OpenBSD like you suggested).

The thing is, the font wouldn't even attempt to open if it was a bug with ghc::filesystem like you suggested, because I am checking every font file for existence before trying to load them, and file_exists() is returning true. This wouldn't be the case if my directory iterator or environment variable was spitting out wrong characters, and I verified by printing to the msys2 terminal the absolute file paths and it correctly showed the Chinese Traditional characters, uncomment this line and see for yourself:

https://github.com/time-killer-games/ImGuiFileDialogCLI/blob/2a5555667f5b81e7d7769200d52c4a07192fdce0/filedialogs.cpp#L206

If you are out of ideas I will see about making an infinitely smaller test case without using a ghc::filesystem and see what happens.

```
git clone https://github.com/time-killer-games/ImGuiFileDialogCLI
cd ImGuiFileDialogCLI
./build.sh
./filedialog --get-directory ''
```

**Comment by ocornut** [comment]
Have you tried doing a fopen() or seeing why the load font function fails? Are you sure your font file is valid?

**Reply by ghost** [reply]
My fonts are valid, it only happens when there are unsupported characters in the path. If i use English-only for example the same files load just fine. I'm using Google's Noto Sans font collection as you guys have recommended. It appears thedmd may have steered me in the right direction, i had a feeling i was doing something wrong, as if this was a bug with imgui you guys have enough non-english users that would be very unlikely. Thank you!

**Reply by ghost** [reply]
Ah, I just noticed the first part of what you said with using fopen. lol trying that now!

Edit: 

I just remembered fopen doesn't support utf-8 so it won't exactly shed light on the situation. _wfopen_s() would work but since the string would need to be converted to a wide string it still might not be obvious whether the string has the correct narrow characters or not.

Edit2:

Ok, oddly enough, both fopen (using sthe string.c_str() as-is) and _wfopen_s (with the wide string variant) worked. So that means fopen() supports narrow characters with utf-8. In either case, it didn't return a nullptr so the functions were able to access the file ok.

**Comment by ghost** [comment]
@ocornut @thedmd I found out the issue. Building with MSVC allows for UTF-8. I was told that this is actually a bug with MinGW.

**Reply by ghost** [reply]
I updated build errors that only happened in MinGW due some changes I made but only tested in MSVC. Run `filedialogs/build.sh` in my repository to demonstrate this issue with UTF-8 so it can be verified again if you want.
